test/test_linalg.py

```diff
@@ -117,6 +117,234 @@ def run_test_case(a, b):
         run_test_case(zero_strided, b)
         run_test_case(a, zero_strided)
 
+    @skipCUDAIfNoMagma
+    @skipCPUIfNoLapack
+    @dtypes(torch.float, torch.double, torch.cfloat, torch.cdouble)
+    def test_linalg_lstsq(self, device, dtype):
+        from torch.testing._internal.common_utils import random_well_conditioned_matrix
+        if self.device_type == 'cpu':
+            drivers = ('gels', 'gelsy', 'gelsd', 'gelss', None)
+        else:
+            drivers = ('gels', None)
+
+        def check_correctness(a, b, sol):
+            sol2 = a.pinverse() @ b
+            self.assertEqual(sol, sol2, atol=1e-5, rtol=1e-5)
+
+        def check_correctness_ref(a, b, res, ref):
+            def apply_if_not_empty(t, f):
+                if t.numel():
+                    return f(t)
+                else:
+                    return t
+
+            def select_if_not_empty(t, i):
+                selected = apply_if_not_empty(t, lambda x: x.select(0, i).numpy())
+                return selected
+
+            m = a.size(-2)
+            n = a.size(-1)
+            nrhs = b.size(-1)
+            a_3d = a.view(-1, m, n)
+            b_3d = b.view(-1, m, nrhs)
+
+            solution_3d = res.solution.view(-1, n, nrhs)
+            residuals_3d = apply_if_not_empty(res.residuals, lambda t: t.view(-1, nrhs))
+            rank_3d = apply_if_not_empty(res.rank, lambda t: t.view(-1))
+            singular_values_3d = apply_if_not_empty(res.singular_values, lambda t: t.view(-1, min(m, n)))
+            batch_size = a_3d.shape[0]
+
+            for i in range(batch_size):
+                sol, residuals, rank, singular_values = ref(
+                    a_3d.select(0, i).numpy(),
+                    b_3d.select(0, i).numpy()
+                )
+                # Singular values are None when lapack_driver='gelsy' in SciPy
+                if singular_values is None:
+                    singular_values = []
+                self.assertEqual(sol, select_if_not_empty(solution_3d, i), atol=1e-5, rtol=1e-5)
+                self.assertEqual(residuals, select_if_not_empty(residuals_3d, i), atol=1e-5, rtol=1e-5)
+                self.assertEqual(rank, select_if_not_empty(rank_3d, i), atol=1e-5, rtol=1e-5)
+                self.assertEqual(singular_values, select_if_not_empty(singular_values_3d, i), atol=1e-5, rtol=1e-5)
+
+        def check_correctness_scipy(a, b, res, driver):
+            if TEST_SCIPY and driver not in (None, 'gels'):
+                import scipy.linalg
+
+                def scipy_ref(a, b):
+                    return scipy.linalg.lstsq(a, b, lapack_driver=driver)
```

Comment:
This test looks very good; should the same {r}cond value be passed to the SciPy and NumPy references, though?

Reply:
They use the same rcond, actually. For numpy i pass `rcond=-1` to use legacy machine precision based approach.

Reply:
Does the test exercise non-default {r}cond vs the references?

Reply:
Also, have we discussed why NumPy (but not SciPy) deprecated that approach? Are we sure we want to follow SciPy's behavior here?

Reply:
> Does the test exercise non-default {r}cond vs the references?

No, it does not

> Also, have we discussed why NumPy (but not SciPy) deprecated that approach? Are we sure we want to follow SciPy's behavior here?

No, we did not, from NumPy:
```
rcond float, optional

    Cut-off ratio for small singular values of a. For the purposes of rank determination, singular values are treated as zero if they are smaller than rcond times the largest singular value of a.

    Changed in version 1.14.0: If not set, a FutureWarning is given. The previous default of -1 will use the machine precision as rcond parameter, the new default will use the machine precision times max(M, N). To silence the warning and use the new default, use rcond=None, to keep using the old behavior, use rcond=-1.

```
I do not know why this additional factor of `max(M, N)` matters, because LAPACK decides that a singular value is zero if `singular_value <= rcond * largest_singular_value`.

Reply:
It'd be nice to make an informed decision ;)

We don't have to block this PR on that decision, however, as long as we put a Note in the documentation that the cond behavior may change in the near future (and then we pick the behavior we like before the 1.9 release). 

Would you update the test (or add a new test) with something that does test non-default rcond values, too? Even a couple fixtures would be nice to validate it's working as expected.